### PR TITLE
OpenVINO Export

### LINKED ISSUE #552: Using YOLOv5 with Neural Compute Stick2
## ❔Question
Hello, I have successfully converted the trained yolov5 model to Intermediate representation to use it with NCS2. However when I load the model to ncs2 it gives wrong results which are all negative values. Loading the same model on CPU runs without any problem and gives correct values. The question is can yolov5 be used in NCS2 and if yes what are the right steps to make it work correctly?
Thanks in advance

### LINKED ISSUE #696: hi! I want to convert yolov5s to openvino. First it will convert to onnx, it is no problem, ans then when i convert onnx to openvino. It occurs problem.
## Additional context
捕获

### LINKED ISSUE #697: convert to openvino
## ❔Question

hi! I want to convert yolov5s to openvino. First it will convert to onnx, it is no problem, ans then when i convert onnx to openvino. It occurs problem.
## Additional context
捕获

### LINKED ISSUE #891: convert yolov5 to openvino
## Before You Start
train your yolov5 model on your own dataset following [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)
- change Activation Function to Relu or Leaky ReLU, because Hardswish is not suported in onnx and openvino.Pre-trained model can be loaded ignoring which Activation Function u use.
https://github.com/ultralytics/yolov5/blob/5e0b90de8f7782b3803fa2886bb824c2336358d0/models/common.py#L26

### before run export.py:
- **changeto torch==1.5.1,torchvision==0.6.1**
- change https://github.com/ultralytics/yolov5/blob/5e0b90de8f7782b3803fa2886bb824c2336358d0/models/yolo.py#L49-L53
```
                c=(y[..., 0:2] * 2. - 0.5 + self.grid[i].to(x[i].device)) * self.stride[i]  # xy
                d=(y[..., 2:4] * 2) ** 2 * self.anchor_grid[i]  # wh
                e=y[..., 4:]
                f=torch.cat((c,d,e),4)
                z.append(f.view(bs, -1, self.no))

        return x if self.training else torch.cat(z, 1)
```
- change https://github.com/ultralytics/yolov5/blob/5e0b90de8f7782b3803fa2886bb824c2336358d0/models/export.py#L31
```
model.model[-1].export = False
```
- set https://github.com/ultralytics/yolov5/blob/5e0b90de8f7782b3803fa2886bb824c2336358d0/models/export.py#L51-L52
to opset_version=10.Because only opset=10 support resize ops.

then you can run export.py to export onnx model.

## install OPENVINO2020R4
you can install following https://bbs.cvmart.net/topics/3117

## convert onnx to openvino through
```bash
python3 /opt/intel/openvino_2020.4.287/deployment_tools/model_optimizer/mo.py \
--input_model {input_dir}/yolov5s.onnx \
--output_dir {output_dir}  \
--input_shape [1,3,480,480]
```
 then you can get openvino model .bin and .xml.

### LINKED ISSUE #970: yolov5s.pt convert to opnevino test time
## ❔Question
I use yolov5s, v3, img size is 640*640, one thread, test openvino(2020R4) time is 1.28s, and pytorch one thread time is 1.2s,   it's too slow.
platform: E5-2699 V3 @2.30GHz.
Is there anything wrong with my test?

### LINKED ISSUE #1193: Openvino Usage
## Additional context
I converted yolov5 to ONNX using export.py, then I cloned the OpenVINO toolkit and installed an inference engine. Then I got the IR files by using mo.py in the cloned OpenVINO toolkit. I managed to run the IR model somehow by using sample codes in OpenVINO. However, now I do not know how to decode the output of the model. Is the procedure the same as in samples of yolov3? Does someone have a script in c++? If yes, I would be more than happy for clues.

### LINKED ISSUE #3063: yolov5 on Intel compute stick2?
## ❔Question
How to run my yolov5 with weights "best.pt" on intel compute stick. 
## Additional context
I have trained yolo on 3200 images with yolo large model.  Then converted ".pt" to ".onnx" with "export.py". Later converted ".onnx" format to ".bin" and ".xml"  format with openvino's model optimizer to run the model on intel compute stick. All the predictions done are incorrect and make no sense. for example: confidence and class label are negative. how to do it in right way. 

I have trained my yolo in this way
python train.py --img 1024 --batch 24 --epochs 100 --data wheat.yaml --cfg models/yolov5l.yaml --name wml

### LINKED ISSUE #5533: Convert Yolov5 to IR Model in Openvino
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello @glenn-jocher et all,

Has anyone ever done the conversion of yolov5 models to IR models in openvino? maybe there is a tutorial i can learn? I want to try deploying yolov5 on Intel NCS2/VPU hardware devices.
thanks

### Additional

_No response_

### PR DESCRIPTION
**Pull request by glenn-jocher**
Adds support for YOLOv5 Intel OpenVINO export.

```python
!git clone https://github.com/ultralytics/yolov5  # clone
%cd yolov5
%pip install -qr requirements.txt onnx openvino-dev  # install INCLUDING onnx and openvino-dev

import torch
from yolov5 import utils
display = utils.notebook_init()  # checks

# Export OpenVINO
!python export.py --include openvino
```

image

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added OpenVINO export support to YOLOv5

### 📊 Key Changes
- 📦 Introduced export functionality for OpenVINO format in `export.py`
- 👾 Updated usage instructions to include `openvino` in the list of exportable formats
- 👁 Added inference command example for OpenVINO models (under development)
- 🛠 Implemented an `export_openvino` function to convert models to OpenVINO format
- 🔄 Modified default `opset` parameter from 14 to 12 for compatibility with OpenVINO
- 📄 Added `openvino-dev` to the requirements in `requirements.txt` (commented out)

### 🎯 Purpose & Impact
- 🌐 Users can now export their YOLOv5 models to OpenVINO, which allows deployment on Intel hardware and accelerators.
- 🚀 This feature broadens the applicability of YOLOv5 models, potentially improving performance on platforms that support OpenVINO.
- 🤝 By commenting out `openvino-dev`, users who require this functionality can easily install it, while those who do not need it avoid unnecessary dependencies.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
OpenVINO PR currently failing with exit code 1. If anyone can help please let me know what's wrong! Thanks!!
Screen Shot 2021-12-21 at 4 46 03 PM

**Comment by glenn-jocher** [comment]
UPDATE: Problem was that OpenVINO export seems to require ONNX Opset <= 12. I've enforced this constraint now and everything seems to be working well :)

Screen Shot 2021-12-21 at 5 03 04 PM

**Comment by glenn-jocher** [comment]
UPDATE2: converted to directory output since OpenVINO creates 3 files. Export directory is i.e. `yolov5s_openvino_model/`:

Screen Shot 2021-12-21 at 5 23 59 PM

**Comment by timmzimm** [comment]
@glenn-jocher, I have problem(
"fatal: Remote branch export/openvino not found in upstream origin"
Could you help please?
